### PR TITLE
MAINT: `stats.bootstrap`: emit `FutureWarning` about broadcasting

### LINKED ISSUE #20850: DOC: stats.bootstrap: improve documentation multidimensional input
Hello! I am writing to describe some errors in the bootstrap function. I am happy to help fix these errors if we agree that they are indeed bugs.

The issue arises when you give the bootstrap a statistic which depends on $X$ and $y$. Say $X \in \mathbb{R}^{n \times d}$ and $y \in \mathbb{R}^n$. (This follows standard scipy conventions, as in [the logistic regression implementation](https://github.com/scikit-learn/scikit-learn/blob/5491dc695dbe2c9bec3452be5f3c409706ff7ee7/sklearn/linear_model/_logistic.py#L1171)). Let's say also that the statistic depends on _both_ $X$ and $y$; for example, let's say the statistic is $\mathbf{1}^{\top}X^{\top}y$. Then we would expect running `bootstrap([X,y], statistic, n_resamples=7, paired=True)` to give us an array of size 7; it should just resample $X$ and $y$ 7 times, and calculate the statistic for each resampling.

However, this is not what happens. In this MWE below, the function _fails to run_. Errors in the resampling protocol are causing this bug. 
1. Firstly, to be consistent with the rest of scipy, we should have the _first_ column index the length of the sample, and the rest of the columns indexx dimensions.
2. Running [`_bootstrap_resample(X, n_resamples=7)`](https://github.com/scipy/scipy/blob/44e4ebaac992fde33f04638b99629d23973cb9b2/scipy/stats/_resampling.py#L625) returns an array with shape `(n, n_resamples, d)`. Running `_bootstrap_resample(y, n_resamples=7)` gives an array of shape `(n_resamples,n)`.
3. Calculating [`statistic(*resampled_data, axis=-1)'](https://github.com/scipy/scipy/blob/44e4ebaac992fde33f04638b99629d23973cb9b2/scipy/stats/_resampling.py#L630) will fail, due to a dimension mismatch. In some other examples, it will fail silently, which can be worse, causing errors that propagate into downstream statistical analyses.

Thank you for your consideration!

```
 import numpy as np
 from scipy.stats import bootstrap
 
 def statistic(x,y):
     v = np.ones((2,))
     vXt = v@x.T
     return vXt@y
 
 X = np.column_stack((0*np.ones((10,)), np.ones((10,))))
 Y = 100*np.ones((10,))
 data = [X,Y]
 n_resamples = 100
 statistic_original = statistic(*data)
 print("Original statistic (working shape): ", statistic_original)
 print("The following line fails due to bad shape!")
 bsd = bootstrap([X,Y], statistic=statistic, n_resamples=15, paired=True).bootstrap_distribution
 print(bsd, bsd.shape)
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-20850

#### What does this implement/fix?
gh-20850 highlighted that `stats.bootstrap` does not explicitly broadcast arrays to the same shape (except along `axis`) as essentially all other reducing `stats` functions do. Since `bootstrap` assumes that the arguments are of the same shape (except along `axis`), I don't think there are any (correct) uses of the function that cannot be accomplished with this restriction, so `bootstrap` should add the broadcasting. There are currently some legitimate uses that *happen* to work even when the shapes are incompatible. To avoid breaking these uses without warning, this PR emits a `FutureWarning` if the shapes do not agree.

#### Additional information
@aangelopoulos I was tempted in this PR to add notes to the documentation about how to use `bootstrap` for samples that contain multidimensional observations. I'd be happy to incorporate such examples here, or you're welcome to submit a separate PR.

### CODE REVIEW

File: scipy/stats/_resampling.py [hunk 1]
```diff
@@ -315,7 +329,17 @@ def bootstrap(data, statistic, *, n_resamples=9999, batch=None,
     Parameters
     ----------
     data : sequence of array-like
-         Each element of data is a sample from an underlying distribution.
+         Each element of `data` is a sample containing scalar observations from an
+         underlying distribution. Elements of `data` must be broadcastable to the
+         same shape (with the possible exception of the dimension specified by `axis`).
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:334]: This would be a great place to refer the reader to a note or example about using `bootstrap` with samples that contain vector observations.

File: scipy/stats/_resampling.py [hunk 2]
```diff
@@ -315,7 +329,17 @@ def bootstrap(data, statistic, *, n_resamples=9999, batch=None,
...
+
+         .. warning:: Beginning in SciPy 1.14.0, `bootstrap` will emit a `FutureWarning`
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_resampling.py:336]: No but there is `versionchanged` which might be useful here.

**Review by tupui — APPROVED** [review]
LGTM otherwise and the added details are nice 👍 

(CI seems partially related)

File: scipy/stats/_resampling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_resampling.py]: OK, changed `warning` to `versionchanged`.

**Review by tupui — APPROVED** [review]
Let's go 🚀
